lib/pubsubstub/subscription.rb

```diff
@@ -0,0 +1,63 @@
+module Pubsubstub
+  class Subscription
+    include Logging
+
+    attr_reader :channels, :connection, :queue, :id
+
+    def initialize(channel_names, connection)
+      @id = Random.rand(2 ** 64)
+      @connection = connection
+      @channels = channel_names.map { |n| Channel.new(n) }
+      @queue = Queue.new
+    end
+
+    def stream(last_event_id)
+      info { "Connecting client ##{id} (#{channels.map(&:name).join(', ')})" }
+      fetch_scrollback(last_event_id)
+      subscribe
```

Comment:
Would you be ok with fetching the scrollback after subscribing? The risk would be that a newly published event would be inserted in the middle of the scrollback and appear twice.

It basically mean we don't guarantee perfect ordering nor perfect unicity, only best effort. I think that's ok, if we wanted to guarantee that we'd need to sort the scrollback events and lock in more places.

Or at least it's ok based on how Shipit uses pubsubstub. If we were to mutate state based on the events it could be problematic unless we were ensuring unicity client side.

Another solution would be to keep the last N (say 1k) ids for each subscriptions, and drop the duplicates. But again I wonder if it's worth it.